homeassistant/components/knx/knx_entity.py

```diff
@@ -43,6 +43,8 @@ async def async_added_to_hass(self) -> None:
         """Store register state change callback."""
         self._device.register_device_updated_cb(self.after_update_callback)
 
+        self.async_on_remove(self._device.shutdown)
+
     async def async_will_remove_from_hass(self) -> None:
         """Disconnect device object when removed."""
         self._device.unregister_device_updated_cb(self.after_update_callback)
```

Comment:
this should be done by `shutdown()` in xknx.
```py

    def shutdown(self) -> None:
        """Prepare for deletion. Remove callbacks and device form Devices vector."""
        self.xknx.devices.remove(self)
        self.device_updated_cbs = []  # <-- here
        ...
```

Reply:
Should we just call `self._device.shutdown()` in `async_remove_from_hass` then?

What is better?

Reply:
I think its more obvious to do it in `remove_form_hass` (or `will_remove_from_hass` - don't know the difference) than to register a callback in `added_to_hass`.